src/hardware_manager.rs

```diff
@@ -0,0 +1,332 @@
+                .spawn(move || NavigationManager::monitor(refresh_interval))
+                .expect("Error: Navigator service: Monitor can't setup thread"),
+        )
+    }
+
+    pub fn init_datalogger(refresh_interval: u64, directory: String, filename: String) {
+        NavigationManager::get_instance().lock().unwrap().datalogger = Some(
+            thread::Builder::new()
+                .name("Navigator service - datalogger".into())
+                .spawn(move || {
+                    NavigationManager::data_logger(refresh_interval, directory, filename)
+                })
+                .expect("Error: Navigator service: Datalogger can't setup thread"),
+        )
+    }
+
+    fn monitor(refresh_interval_us: u64) {
+        log::info!("Monitor: Started");
+        loop {
+            let time_start = std::time::Instant::now();
+
+            let mut lock = Self::get_instance().lock().unwrap();
+
+            let reading = navigator_rs::SensorData {
+                adc: {
+                    if refresh_interval_us < 10000 {
+                        navigator_rs::ADCData { channel: [0.0; 4] }
+                    } else {
+                        lock.navigator.read_adc_all()
+                    }
+                },
+                temperature: lock.navigator.read_temperature(),
+                pressure: lock.navigator.read_pressure(),
+                accelerometer: lock.navigator.read_accel(),
+                magnetometer: lock.navigator.read_mag(),
+                gyro: lock.navigator.read_gyro(),
+            };
+
+            drop(lock);
+
+            *DATA.write().unwrap() = Data { state: reading };
+
+            let time_elapsed = time_start.elapsed().as_micros() as u64;
+
+            if time_elapsed > (refresh_interval_us * 2) {
+                log::info!("Monitor: Something went wrong, measurements not concluded with reading interval {refresh_interval_us} us, time elapsed: {time_elapsed} us");
+                NavigationManager::websocket_broadcast();
+                continue;
+            }
+
+            NavigationManager::websocket_broadcast();
+
+            let wait = refresh_interval_us as i64 - time_elapsed as i64;
+
+            if wait > 0 {
+                thread::sleep(std::time::Duration::from_micros(wait as u64));
+            }
+        }
+    }
+
+    fn data_logger(refresh_interval: u64, directory: String, filename: String) {
+        let file_path = format!("{}/{}", directory, filename);
+        let mut logger = DataLogger::new(&file_path).expect("Failed to create/open CSV file");
+
+        // Just let monitor run before
+        thread::sleep(std::time::Duration::from_millis(500));
+
+        log::info!("Datalogger started");
+
+        loop {
+            let reading = DATA.read().unwrap().state;
+
+            logger.log_data(&reading).expect("Failed to log data");
+
+            thread::sleep(std::time::Duration::from_micros(refresh_interval));
```

Comment:
```suggestion
            thread::sleep(std::time::Duration::from_micros(refresh_interval));

            let reading = DATA.read().unwrap().state;

            logger.log_data(&reading).expect("Failed to log data");
```

It's better to sleep on the beginning as a good practice to avoid spinlock

Reply:
Ok